x-pack/plugins/security/server/session_management/session_management_service.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Observable, Subscription } from 'rxjs';
+import { HttpServiceSetup, ILegacyClusterClient, Logger } from '../../../../../src/core/server';
+import { TaskManagerSetupContract, TaskManagerStartContract } from '../../../task_manager/server';
+import { ConfigType } from '../config';
+import { OnlineStatusRetryScheduler } from '../elasticsearch';
+import { SessionCookie } from './session_cookie';
+import { SessionIndex, SESSION_INDEX_TEMPLATE_VERSION } from './session_index';
+import { Session } from './session';
+
+export interface SessionManagementServiceSetupParams {
+  readonly http: Pick<HttpServiceSetup, 'basePath' | 'createCookieSessionStorageFactory'>;
+  readonly config: ConfigType;
+  readonly clusterClient: ILegacyClusterClient;
+  readonly kibanaIndexName: string;
+  readonly taskManager: TaskManagerSetupContract;
+}
+
+export interface SessionManagementServiceStartParams {
+  readonly online$: Observable<OnlineStatusRetryScheduler>;
+  readonly taskManager: TaskManagerStartContract;
+}
+
+export interface SessionManagementServiceSetup {
+  readonly session: Session;
+}
+
+/**
+ * Name of the task that is periodically run and performs session index cleanup.
+ */
+export const SESSION_INDEX_CLEANUP_TASK_NAME = 'session_cleanup';
+
+/**
+ * Service responsible for the user session management.
+ */
+export class SessionManagementService {
+  private statusSubscription?: Subscription;
+  private sessionIndex!: SessionIndex;
+  private config!: ConfigType;
+
+  constructor(private readonly logger: Logger) {}
+
+  setup({
+    config,
+    clusterClient,
+    http,
+    kibanaIndexName,
+    taskManager,
+  }: SessionManagementServiceSetupParams): SessionManagementServiceSetup {
+    this.config = config;
+
+    const sessionCookie = new SessionCookie({
+      config,
+      createCookieSessionStorageFactory: http.createCookieSessionStorageFactory,
+      serverBasePath: http.basePath.serverBasePath || '/',
+      logger: this.logger.get('cookie'),
+    });
+
+    this.sessionIndex = new SessionIndex({
+      config,
+      clusterClient,
+      indexName: `${kibanaIndexName}_security_session_${SESSION_INDEX_TEMPLATE_VERSION}`,
```

Comment:
nit: it feels "off" that we need to import this constant from the `SessionIndex` module in order to create the correct index name when constructing an instance of `SessionIndex`. What if `SessionIndex` just accepted the current kibana index name, and it constructed the final session index name itself?

Reply:
> nit: it feels "off" that we need to import this constant from the SessionIndex module in order to create the correct index name when constructing an instance of SessionIndex. What if SessionIndex just accepted the current kibana index name, and it constructed the final session index name itself?

I agree, I had the same concern and was basically choosing between this and spreading knowledge about Kibana index in more places. I don't like both (and as you might have already noticed our legacy multi-tenancy support altogether :laughing: ), but with your vote we have more proponents for the first option - I'll change it!